### ISSUE TITLE
can not receive different outputs from mGENRE.sample using dropout in train mode and different seeds

### ISSUE THREAD

**Issue by SergeyPetrakov** [facebookresearch/GENRE#83]
Thank you for your work! 
I am trying to receive different scores from mGENRE model using dropout and switching model to the train mode
1) I changed in fairseq_model.py train function in order to be able to train mode
```
def train(mode=True):
    self.train()
```
    
2) then I turned on dropout and switched to train mode
```
model_mGENRE_mcdropout = mGENRE.from_pretrained("fairseq_multilingual_entity_disambiguation",
                                                dropout = 0.9, attention_dropout = 0.9)
model_mGENRE_mcdropout.train()
```
3) I checked that dropout is working (it really works)
```
model_mGENRE_mcdropout.models[0].encoder.dropout_module.forward(torch.ones(100))
```

But I cannot receive different scores using the following command 
```
import random

for i in range(3):
    random.seed(i)
    torch.manual_seed(i)
    np.random.seed(i)
    print(model_mGENRE_mcdropout.sample(["[START] Caesar [END] is well known Roman imperor"],
                                  beam = 3,
                                  prefix_allowed_tokens_fn=lambda batch_id, sent: [
                                      e for e in trie.get(sent.tolist())
                                      if e < len(model_mGENRE_mcdropout.task.target_dictionary)
                                  ],
                                  text_to_id=lambda x: max(lang_title2wikidataID[tuple(reversed(x.split(" >> ")))], key=lambda y: int(y[1:])),
                                  marginalize=True,
                                  verbose = False,
                                  seed = i))
    print("\n")
```
output is just 
[[{'id': 'Q1048', 'texts': ['Caesar >> en', 'Julius Caesar >> en'], 'scores': tensor([-0.3048, -0.3708]), 'score': tensor(-0.0202)}, {'id': 'Q188751', 'texts': ['Caesar (title) >> en'], 'scores': tensor([-0.3666]), 'score': tensor(-0.9700)}]]

[[{'id': 'Q1048', 'texts': ['Caesar >> en', 'Julius Caesar >> en'], 'scores': tensor([-0.3048, -0.3708]), 'score': tensor(-0.0202)}, {'id': 'Q188751', 'texts': ['Caesar (title) >> en'], 'scores': tensor([-0.3666]), 'score': tensor(-0.9700)}]]

[[{'id': 'Q1048', 'texts': ['Caesar >> en', 'Julius Caesar >> en'], 'scores': tensor([-0.3048, -0.3708]), 'score': tensor(-0.0202)}, {'id': 'Q188751', 'texts': ['Caesar (title) >> en'], 'scores': tensor([-0.3666]), 'score': tensor(-0.9700)}]]

Could you please tell me how can I change the code in order to receive different scores?

**Comment by nicola-decao** [comment]
Honestly, I have no idea. Looks like you need to modify the `fairseq` behaviour, i.e., probably in `fairseq` a `.eval()` is called somewhere.

**Comment by SergeyPetrakov** [comment]
Thank you for such a fast answer, I will try to find ```.eval()``` somewhere in ```fairseq```
